src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameParser.cs

```diff
@@ -0,0 +1,455 @@
+        ///   * MyClass.MyMethod([|$$string[]|] s)
+        ///   * [|$$MyClass`1.MyMethod|](string s)
+        ///   * [|$$MyClass.MyMethod|][T](T t)
+        /// </code>
+        /// 
+        /// </summary>
+        private ParseResult<StackFrameNameNode> TryParseNameNode(bool scanAtTrivia)
+        {
+            var currentIdentifer = _lexer.TryScanIdentifier(scanAtTrivia: scanAtTrivia, scanLeadingWhitespace: true, scanTrailingWhitespace: false);
+            if (!currentIdentifer.HasValue)
+            {
+                // If we can't parse an identifier, no need to abort the parser entirely here. 
+                // The caller should determine if an identifier is required or not.
+                return ParseResult<StackFrameNameNode>.Empty;
+            }
+
+            var identifierParseResult = TryScanGenericTypeIdentifier(currentIdentifer.Value);
+            if (!identifierParseResult.Success)
+            {
+                return ParseResult<StackFrameNameNode>.Abort;
+            }
+
+            RoslynDebug.AssertNotNull(identifierParseResult.Value);
+            var lhs = identifierParseResult.Value;
+
+            var parseResult = TryParseQualifiedName(lhs);
+            if (!parseResult.Success)
+            {
+                return ParseResult<StackFrameNameNode>.Abort;
+            }
+
+            var memberAccess = parseResult.Value;
+            if (memberAccess is null)
+            {
+                Debug.Assert(lhs is StackFrameSimpleNameNode);
+                return new(lhs);
+            }
+
+            while (true)
+            {
+                parseResult = TryParseQualifiedName(memberAccess);
+                if (!parseResult.Success)
+                {
+                    return ParseResult<StackFrameNameNode>.Abort;
+                }
+
+                var newMemberAccess = parseResult.Value;
+                if (newMemberAccess is null)
+                {
+                    Debug.Assert(memberAccess is StackFrameQualifiedNameNode);
+                    return new(memberAccess);
+                }
+
+                memberAccess = newMemberAccess;
+            }
+        }
+
+        /// <summary>
+        /// Given an existing left hand side node or token, which can either be 
+        /// an <see cref="StackFrameKind.IdentifierToken"/> or <see cref="StackFrameQualifiedNameNode"/>
+        /// </summary>
+        private ParseResult<StackFrameQualifiedNameNode> TryParseQualifiedName(StackFrameNameNode lhs)
+        {
+            if (!_lexer.ScanCurrentCharAsTokenIfMatch(StackFrameKind.DotToken, out var dotToken))
+            {
+                return ParseResult<StackFrameQualifiedNameNode>.Empty;
+            }
+
+            var identifier = _lexer.TryScanIdentifier();
+            if (!identifier.HasValue)
+            {
+                return ParseResult<StackFrameQualifiedNameNode>.Abort;
+            }
+
+            var (success, rhs) = TryScanGenericTypeIdentifier(identifier.Value);
```

Comment:
merge this and the above into TryScanSimpleNameNode.  You have this pattern in a few places and it will clean things up (and help with the static typeing of things).

Reply:
Not sure I follow exactly. The split now makes things easy to read in `TryParseRequiredNameNode`. Is this an old comment?